projects/matlib/Mat.h

```diff
@@ -36,58 +42,110 @@ class Mat {
             return iterator(*this, 0);
         }
         iterator end(){
-            return iterator(*this, strides[0]*columns() + strides[1]*rows());
+            if(ndims == 2){
+                return iterator(*this, strides[0]*columns() + strides[1]*rows());
+            }
+            else{
+                return iterator(*this, strides[0]*columns());
+            }
+        }
+        size_type size() const{
+            if(ndims == 0) return 0;
+            size_type result = dims[0];
+            for(int i = 1; i < ndims; i++){
+                result *= dims[i];
+            }
+            return result;
         }
         size_type rows() const{
-            return this->dims[0];
+            errorCheck(ndims < 2, "1d matrix has no rows");
+            return this->dims[ndims - 2];
         }
         size_type columns() const{
-            return this->dims[1];
+            return this->dims[ndims - 1];
         }
-        size_type size() const{
-            return rows()*columns();
+        bool inbounds(size_type a, size_type b){
+            if(a >= 0 && a < columns() && b >= 0 && b < rows()) return true;
+            else return false;
         }
-        Mat(size_type a = 1, size_type b = 1){
+        bool isContiguous(){
+            if(strides[0] != 1) return false;
+            for(int i = 1; i < ndims; i++){
+                strides[i] != dims[i];
+            }
+        }
+        Mat(size_type a = 1){
             refCount = new int64_t;
             *refCount = 1;
-            memory = new Type[a*b];
+            ndims = 1;
+            dims = new size_type[ndims];
+            dims[0] = a;
+            strides = new size_type[ndims];
+            strides[0] = 1;
+            memory = new Type[a];
             data = memory;
+        }
+        Mat(size_type a, size_type b){
```

Comment:
signature
Mat(Type*, int64_t*)

usage:
double a[4] = {1,2,3,4};
int64_t counter=1;
Mat wrapper = Mat(a, &counter);

Reply:
additional signature
Mat(Type*)
that automatically makes a reference counter that starts at 1

Reply:
What do if original container deletes data?